Improve doc for torch.add and add doc for torch.sub.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25115 Add in-place operator rsub_.
* **#25114 Improve doc for torch.add and add doc for torch.sub.**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Add doc for torch.sub.

xuhdev committed:
Update on "Add doc for torch.sub."

gchanan reviewed (changes requested):
Don't document this, it's considered deprecated, see https://github.com/pytorch/pytorch/blob/master/tools/autograd/deprecated.yaml.

vishwakftw commented on torch/_torch_docs.py:
```diff
@@ -210,6 +210,75 @@ def merge_dicts(*dicts):
+
+Args:
+    {input}
+    value (Number): the number to be added to each element of :attr:`input`
```
`value` is not an argument for this variant I suppose.

vishwakftw commented on torch/_torch_docs.py:
```diff
@@ -210,6 +210,75 @@ def merge_dicts(*dicts):
             [ -8.9902,  -8.3667,  -7.3925,  -7.6147]])
 """.format(**common_args))
 
+add_docstr(torch.sub,
```
It would also be nice if this could be placed in alphabetical order within the file.

xuhdev commented:
> Don't document this, it's considered deprecated, see https://github.com/pytorch/pytorch/blob/master/tools/autograd/deprecated.yaml.

@gchanan I'm a bit confused. What is the replacement? How would one now, for example, do the fma operation (as `add` is also deprecated)?

xuhdev committed:
Update on "Add doc for torch.sub."

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

gchanan commented on torch/_torch_docs.py:
```diff
@@ -181,15 +155,15 @@ def merge_dicts(*dicts):
 
-If :attr:`other` is of type FloatTensor or DoubleTensor, :attr:`alpha` must be
-a real number, otherwise it should be an integer.
+If the result of adding :attr:`self` and :attr:`other` would not have a
```
this isn't right after @nairbv's changes (also with sub below).  More accurate is something like "the category (floating point, integer, boolean)" of alpha must match that of the result of addition between `input` and `other`.

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -181,15 +155,15 @@ def merge_dicts(*dicts):
-If :attr:`other` is of type FloatTensor or DoubleTensor, :attr:`alpha` must be
-a real number, otherwise it should be an integer.
+The category (floating point, integer, or Boolean) of :attr:`alpha` must match
+that of the result of addition between :attr:`input` and :attr:`other`.
```
This isn't true. the following (integer alpha, float result) works for me on master:
```
import torch

x = torch.randn(3, 3)
y = torch.randn(3, 3)
torch.add(x, y, alpha=1)
```

zou3519 replied:
I would change this to say the following:

```
If the result of the addition between :attr:`input` and :attr:`other` is:
- floating point, :attr:`alpha` can be `float` or `int`
- integral, :attr:`alpha` must be `int`
- boolean, :attr:`alpha` can be `int` or `bool.
```

xuhdev replied:
Thanks, I reworded a bit and updated this PR.

zou3519 reviewed (changes requested):
"the category (floating point, integer, or Boolean) of :attr:`alpha` must match
a real number, otherwise it should be an integer.	that of the result of addition between :attr:`input` and :attr:`other`." is wrong, let's reword it.

xuhdev committed:
Update on "Improve doc for torch.add and add doc for torch.sub."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/25114

[ghstack-poisoned]

xuhdev commented:
@gchanan I don't think so -- @nairv's PR is for the 1.3 branch, but this is for master

xuhdev commented:
Of course, if you can cherry-pick the add and sub part of #27501 into master, I think this can be closed

nairbv commented:
> Of course, if you can cherry-pick the add and sub part of #27501 into master, I think this can be closed

My PR on the 1.3 branch conflicts with master, so probably easier to land this one than to cherry pick that one.

xuhdev commented:
Closed in favor of #28118, which also adds doc for `mul` and updates format starting from #27571
